### PR TITLE
fix: vue2.7 SFC compiler.parse error v6

### LINKED ISSUE #765: Vue SFC TS check was broken on vue2.7
## Current behavior

https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/blob/main/src/typescript/extension/vue/type-script-vue-extension.ts#L113

vue2.7 hit `isVueTemplateCompilerV3`,but there is no `descriptor` on `parsed`

```typescript
const parsed = compiler.parse(vueSourceText);

if (parsed.descriptor && parsed.descriptor.script) {
  const scriptV3 = parsed.descriptor.script;
```
`parsed` shows like following codes
```
 {
  source: undefined,
  filename: 'anonymous.vue',
  template: null,
  script: null,
  scriptSetup: null,
  styles: [],
  customBlocks: [],
  cssVars: [],
  errors: [],
  shouldForceReload: [Function (anonymous)]
}

```

So that any ts check on vue will fail.

```
export { default as FilteredSearchToken } from './FilteredSearchToken.vue';
```

image

Error shows:

TS2322: Type '{}' is not assignable to type 'VueConstructor<Vue<Record<string, any>, Record<string, any>, never, never, never, never, (event: string, ...args: any[]) => Vue<Record<string, any>, Record<string, any>, ... 4 more ..., ...>>>'.

## Expected behavior

Checker should work fine.

## Steps to reproduce the issue

```bash
yarn serve
```

## Issue reproduction repository

https://github.com/iceprosurface/vue2.7-tsx-error

## Environment
 - fork-ts-checker-webpack-plugin@7.2.12 & fork-ts-checker-webpack-plugin@6.5.2
 - typescript@4.7.4
 - eslint@8.20.0
 - webpack@5.69.1
 - osx 12.3.1

### PR DESCRIPTION
**Pull request by iceprosurface**
see #765

### CODE REVIEW

**Review by piotr-oles — COMMENTED** [review]
Thanks for the PR - could you add an e2e test case to this file: https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/blob/main/test/e2e/type-script-vue-extension.spec.ts for vue 2.7?

File: src/typescript-reporter/extension/vue/TypeScriptVueExtension.ts [hunk 1]
```diff
@@ -23,17 +24,24 @@ function createTypeScriptVueExtension(
...
     return typeof (compiler as VueTemplateCompilerV2).parseComponent === 'function';
   }
+  function isVueTemplateCompilerV2_7(
+    compiler: VueTemplateCompilerV3 | VueTemplateCompilerV2_7
+  ): compiler is VueTemplateCompilerV2_7 {
+    // FIXME: this is a hack to detect VueTemplateCompilerV2_7,should use version to detect.
+    return (compiler as VueTemplateCompilerV2_7).parse.length === 1;
+  }
 
   function isVueTemplateCompilerV3(
```
Comment (piotr-oles) [inline src/typescript-reporter/extension/vue/TypeScriptVueExtension.ts]: could you rename this function to `isVueTemplateCompilerV2_7OrV3`?

**Comment by iceprosurface** [comment]
I think we should use vue's semantic version to check whether vue is vue2.7 | vue3 | vue2.

`VueTemplateCompilerV2_7` & `isVueTemplateCompilerV2_7OrV3` & `isVueTemplateCompilerV2` cannot be able to a safely check vue versions.

From: https://github.com/vuejs/vue/issues/12679#issuecomment-1189778234

**Comment by piotr-oles** [comment]
Feel free to change that. This would not prevent this issue from happening - they introduced breaking change to their API without a major bump ;)

**Comment by iceprosurface** [comment]
I don‘t know why e2e was fail on github action 

it works well on my mac.

image


ohh, There is so many lock file changes,may i append all lock file changes?
image

**Comment by piotr-oles** [comment]
Yes, the locks file in git is not a great DX - that's why I changed it in the v7. You can commit these lock files.

**Comment by rs3d** [comment]
Hi, I've noticed also the missing type-checks for vue@2.7, which is unfortunate because the prod-build is not stopping anymore if there are any type errors. Is this PR going to be merged soon-ish?

**Comment by mohanyin** [comment]
Thanks @iceprosurface for the PR! I've been trying to figure out this bug all morning.

While we're waiting for this to get merged, has anyone figured out a workaround?

**Comment by piotr-oles** [comment]
I'm dropping support for Vue.js - there is no clean API in TypeScript to support it and I don't have time to maintain it as current workaround is hacky and buggy :/

**Pull request closed without merging**